cpp/src/copying/scatter.cu

```diff
@@ -0,0 +1,469 @@
+    }
+
+    return output;
+  }
+
+  template <typename T, std::enable_if_t<not std::is_integral<T>::value
+      or std::is_same<T, bool8>::value>* = nullptr>
+  std::vector<std::unique_ptr<table>> operator()(
+      table_view const& input, column_view const& partition_map,
+      rmm::mr::device_memory_resource* mr, cudaStream_t stream)
+  {
+    CUDF_FAIL("Partition map column must be an integral, non-boolean type");
+  }
+};
+
+}  // namespace
+
+std::unique_ptr<table> scatter(
+    table_view const& source, column_view const& scatter_map,
+    table_view const& target, bool check_bounds,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  CUDF_EXPECTS(source.num_columns() == target.num_columns(),
+    "Number of columns in source and target not equal");
+  CUDF_EXPECTS(scatter_map.size() <= source.num_rows(),
+    "Size of scatter map must be equal to or less than source rows");
+  CUDF_EXPECTS(std::equal(source.begin(), source.end(), target.begin(),
+    [](auto const& col1, auto const& col2) {
+      return col1.type().id() == col2.type().id();
+    }), "Column types do not match between source and target");
+  CUDF_EXPECTS(scatter_map.has_nulls() == false, "Scatter map contains nulls");
+
+  if (scatter_map.size() == 0) {
+    return std::make_unique<table>(target, stream, mr);
+  }
+
+  // First dispatch for scatter map index type
+  return type_dispatcher(scatter_map.type(), scatter_impl{}, source,
+    scatter_map, target, check_bounds, mr, stream);
+}
+
+std::unique_ptr<table> scatter(
+    std::vector<std::unique_ptr<scalar>> const& source, column_view const& indices,
+    table_view const& target, bool check_bounds,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  CUDF_EXPECTS(source.size() == static_cast<size_t>(target.num_columns()),
+    "Number of columns in source and target not equal");
+  CUDF_EXPECTS(std::equal(source.begin(), source.end(), target.begin(),
+    [](auto const& scalar, auto const& col) {
+      return scalar->type().id() == col.type().id();
+    }), "Column types do not match between source and target");
+  CUDF_EXPECTS(indices.has_nulls() == false, "indices contains nulls");
+
+  if (indices.size() == 0) {
+    return std::make_unique<table>(target, stream, mr);
+  }
+
+  // First dispatch for scatter index type
+  return type_dispatcher(indices.type(), scatter_scalar_impl{}, source,
+    indices, target, check_bounds, mr, stream);
+}
+
+std::vector<std::unique_ptr<table>> scatter_to_tables(
+    table_view const& input, column_view const& partition_map,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+{
+  CUDF_EXPECTS(partition_map.size() <= input.num_rows(), "scatter map larger than input");
+  CUDF_EXPECTS(partition_map.has_nulls() == false, "scatter map contains nulls");
+
+  if (partition_map.size() == 0 || input.num_rows() == 0) {
+    return std::vector<std::unique_ptr<table>>{};
+  }
+
+  // First dispatch for scatter index type
```

Comment:
Is there a second dispatch?

Reply:
This may have been a slip of copy-paste... But yes! The first dispatch here is over the partition/scatter map column type and the second dispatch for column data type happens inside the call to gather. I'll try to make that more clear

Reply:
Added comment to clarify where the second dispatch is